stan/math/prim/mat/fun/gp_exponential_cov.hpp

```diff
@@ -0,0 +1,253 @@
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = x.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_exponential_cov", "x", x[n]);
+
+  check_positive("gp_exponential_cov", "marginal variance", sigma);
+  check_not_nan("gp_exponential_cov", "marginal variance", sigma);
+
+  check_positive("gp_exponential_cov", "length-scale", length_scale);
+  check_not_nan("gp_exponential_cov", "length-scale", length_scale);
+
+  Eigen::Matrix<typename stan::return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l neg_inv_l = -1.0 / length_scale;
+
+  for (size_t i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      cov(i, j) = sigma_sq * exp(neg_inv_l * squared_distance(x[i], x[j]));
+      cov(j, i) = cov(i, j);
+    }
+  }
+  cov(x_size - 1, x_size - 1) = sigma_sq;
+  return cov;
+}
+
+/** Returns a Matern exponential covariance matrix with one input vector
+ *  with automatic relevance determination (ARD) priors
+ *
+ * \f[ k(x, x') = \sigma^2 exp(-\sum_{k=1}^K\frac{d(x, x')}{l_k}) \f]
+ *
+ * where \f$ d(x, x') \f$ is the squared distance, or dot product.
+ *
+ * See Rausmussen & Williams et al 2006 Chapter 4.
+ *
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ *
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename stan::return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_exponential_cov(const std::vector<T_x> &x, const T_s &sigma,
+                   const std::vector<T_l> &length_scale) {
+  using stan::math::square;
+  using stan::math::squared_distance;
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = x.size();
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_exponential_cov", "x", x[n]);
+
+  check_positive("gp_exponential_cov", "marginal variance", sigma);
+  check_not_nan("gp_exponential_cov", "marginal variance", sigma);
+
+  check_positive("gp_exponential_cov", "length-scale", length_scale);
+  check_not_nan("gp_exponential_cov", "length-scale", length_scale);
+
+  Eigen::Matrix<typename stan::return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l temp;
+
+  for (size_t i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      temp = 0;
+      for (size_t k = 0; k < l_size; ++k)
+        temp += squared_distance(x[i], x[j]) / length_scale[k];
```

Comment:
This does the right thing, but is written inefficiently.

That `squared_distance()` call could be pretty expensive. And divisions are expensive. The way this is written, there are k x `squared_distance()` calls, k x divisions, and k x sums.

I'd go with: `squared_distance(x[i], x[j]) / sum(length_scale);` That should be 1 x `squared_distance()`, 1 x division, k x sum.

Reply:
and you can get rid of the temp and put this into line 121 directly.